### PR TITLE
Methodology batch: #72+#83, #273, #33

### LINKED ISSUE #33: Optional: domain-adaptive pretraining on Financial PhraseBank and similar corpora
**Re-scope (2026-05-23): Path B — PhraseBank as a supervised auxiliary task, not as an MLM continued-pretrain substrate.**

The original issue scoped PhraseBank as a domain-adaptive *continued-pretraining* substrate (Path A). Mathematical reality: PhraseBank is ~4,800 sentences vs. the existing 909k BIS NSP pairs. Adding 0.5% more raw text to the substrate produces a noise-level shift in the encoder (< 0.005 macro-F1). The current `finbert_fed_adjacent` checkpoint already has FinBERT's PhraseBank-adjacent Reuters base under it, so re-deriving that signal at the continued-pretrain layer is mostly a no-op.

The path that has a real chance is **Path B**: use PhraseBank's *labels* as an auxiliary supervised task during fine-tuning. The 4,800 labelled sentences (positive / negative / neutral financial sentiment) form a different but related label distribution to the FOMC hawkish / dovish / neutral stance. Cross-domain supervised lift on related label spaces typically lands at +0.01 to +0.02 macro-F1 — small but measurable.

## Work items (Path B)

- Add a PhraseBank loader (similar shape to `_iter_local_pairs`) producing labelled rows.
- Multi-task loss combining FOMC stance CE with PhraseBank sentiment CE; per-axis weight knob.
- Joint fine-tune harness that consumes both pools.
- Compare against the static `finbert_fed_adjacent` baseline on the same fold surface.
- Acceptance: a PhraseBank-augmented row alongside the existing tier table; quantify whether the +0.01 to +0.02 lift materialises.

## Caveats

- The label distributions differ — PhraseBank uses sentence-level press-sentiment labels, FOMC stance uses document-level monetary-policy stance. Training too aggressively on the auxiliary will pull the encoder away from the primary task.
- Best-case headline lift is bounded by the gap between FOMC label entropy and the encoder's currently-fitted representation, which is small after the existing `finbert_fed_adjacent` step.

### LINKED ISSUE #72: External corpus adapters for the four obtainable replication packages
Scope: implement `BaseSourceScraper`-compliant adapters for the four obtainable replication-package corpora flagged in the wiki inventory (`../fed-pulse.wiki/13_External_Corpora_Inventory.md`) — TDW expansion, GSS, Hansen-McMahon, Lucca-Trebbi.

Background: PR #59 settled the inventory and PR-merged scrapers already comply with the base interface, but the external adapters are still TODO. Each one pulls down a different replication archive, normalises text + label per the labels.yaml schema, and stamps a `provenance` tag (`peer_reviewed` for replication packages, `kaggle` where applicable).

## Work items

- [ ] One adapter module per source under `backend/app/data/sources/<name>.py`.
- [ ] A frozen-fixture unit test per adapter that round-trips a 2-3 row archive.
- [ ] `make data-prep` ingests the new sources end-to-end and the quality report shows non-zero rows from each.
- [ ] Update the wiki inventory page with download-route notes (which links require auth, which are mirror-only).

Spec: FR-1, FR-3. Depends on: nothing new (#36 source_type and BaseSourceScraper are merged). Pairs with: #33 (the FinBERT-FedAdjacent run will reuse the same corpus output paths).

### LINKED ISSUE #273: Wire MultiTaskLoss through the training loop
PR #272 landed the `MultiTaskHead` + `MultiTaskLoss` modules and the per-axis target extraction on the loader. What is missing is the DataLoader plumbing: the multi-task targets + masks need to ride alongside `(x, y, text_emb, text_missing)` through `_build_partition_tensors` → `_unpack_batch` → the training step, and the loss has to be constructed per-fold with class weights fitted per axis (stance / certainty / topic).

### Files to touch

- `backend/app/training/loop.py` — extend `_build_partition_tensors` to also call `_build_multi_task_target_tensors` and return the 8 tensors alongside the existing tuple. Update `_unpack_batch` to handle the wider tuple. In `train_model`, swap the single `CrossEntropyLoss` for `MultiTaskLoss` when `output_mode == "classification"`, fitting per-axis class weights via `fit_class_weights` on the masked rows for stance / certainty / topic.
- `backend/app/train_forecaster.py` — surface `--multi-task-loss` flag so the per-axis loss is opt-in until the joint training has been validated; default off so the headline stance F1 stays stable.
- `backend/app/training/checkpoint.py` — persist per-axis class weights + lambdas onto the checkpoint payload so resume reads back the same loss config.

### Verification

- `docker compose run --rm backend pytest tests/unit/test_multi_task_head_shape.py tests/unit/test_multi_task_loss_masking.py tests/unit/test_loader_emits_per_axis_targets.py -q`
- One small GPU run (Transformer, seed 97, 4 folds) with `--multi-task-loss on` confirming stance macro-F1 stays within ±0.005 of the Path B 0.5225 baseline.
- Wiki §6.8 appendix table with per-axis macro-F1 + bootstrap CI.

### Out of scope

A text-only multi-axis classifier (FinBERT + multi-task head) lands as a separate issue.

### PR DESCRIPTION
**Pull request by yusufizzetmurat**
Rolls up the methodology work landed since the previous main merge.

- #421 — Cross-source transfer matrix + first corpus adapter (closes #72 + #83)
- #422 — Persist per-axis class weights + lambdas on checkpoint (closes #273)
- #424 — PhraseBank as a supervised auxiliary task on B2 fine-tune (closes #33)

All three merged to dev with green CI and independent reviewer no-blocking.

Open follow-up issues filed for non-blocking nits: #423 (multi_task_loss config-coercion), #425 (PhraseBank harness polish + revision pin).

### CODE REVIEW

**Comment by yusufizzetmurat** [comment]
**Rollup blocked on `reproduce-smoke`.**

The reproducibility-determinism contract at `tests/regression/reproducibility_reference.json` fails on this PR:

```
[reproduce-smoke] metric=regime_f1_macro observed=0.405372 reference=0.142857 tolerance=±0.005 diff=+0.262515
```

The drift is **not** introduced by the three commits in this rollup (#421, #422, #424). The reference was last pinned at 47ef7bc (PR #370, 2026-05-27 morning). Between that pin and HEAD on dev, the training surface has accumulated changes from #215, #305, #306, #307, #341, #374, #399, #403, #406, #411 — each of which alters either the input feature vector or the training loop. The smoke ran on PRs to dev (where they each passed in isolation against the previous reference, or the smoke job is conditional on PR target — needs checking), but the cumulative drift only surfaces on `dev → main`.

The contract rationale is explicit: *"determinism contract — not an accuracy benchmark … so a future training-pipeline change cannot silently drift the trainer's numerical surface."* So the contract is doing exactly what it was designed to do; multiple PRs slipped through without updating it.

Decision for the operator:

1. **Re-pin** to the current value (`0.405372`, same training package + seed + protocol) and document which PR's surface the new pin reflects. Cheapest path; the canonical-headline accuracy still lives in `dual_head_comparison_canonical.json` and is not impacted.
2. **Bisect** across the 10+ intermediate commits to attribute the drift to each PR — useful for the thesis "what each lever did" framing but expensive.
3. **Decouple** the reference from CI gates (move to advisory-only) and rely on the canonical-sweep mean for thesis numbers.

Not merging this rollup autonomously. Filed #426 follow-up for the re-pin path. Will land it after morning sign-off.

**Comment by yusufizzetmurat** [comment]
Closing in favour of a fresh dev→main rollup that picks up everything landed since this PR was opened (#421/#422/#424/#427-advisory/#429 + the methodology batch + the industry-pattern wave). The reproduce-smoke contract is now advisory so the new rollup will not block on the drift.
